ratis-server/src/main/java/org/apache/ratis/server/impl/TransferLeadership.java

```diff
@@ -134,12 +134,11 @@ boolean isSteppingDown() {
   void onFollowerAppendEntriesReply(LeaderStateImpl leaderState, FollowerInfo follower) {
     // If the transferee has just append some entries and becomes up-to-date,
     // send StartLeaderElection to it
-    if (getTransferee().filter(t -> t.equals(follower.getId())).isPresent()) {
-      final String error = leaderState.sendStartLeaderElection(follower, leaderState.getLastEntry());
-      if (error == null) {
-        LOG.info("{}: sent StartLeaderElection to transferee {} after received AppendEntriesResponse",
-            server.getMemberId(), follower.getId());
-      }
+    if (getTransferee().filter(t -> t.equals(follower.getId())).isPresent()
+        && LeaderStateImpl.isFollowerUpToDate(follower, leaderState.getLastEntry())) {
+      LOG.info("{}: sending StartLeaderElection to transferee {} after received AppendEntriesResponse",
+          server.getMemberId(), follower.getId());
+      leaderState.sendStartLeaderElection(follower.getId(), leaderState.getLastEntry());
```

Comment:
- It should reuse `leaderState.getLastEntry()`.

- Check return value before LOG
```java
    if (!getTransferee().filter(t -> t.equals(follower.getId())).isPresent()) {
      return;
    }
    final TermIndex lastEntry = leaderState.getLastEntry();
    if (LeaderStateImpl.isFollowerUpToDate(follower, lastEntry)) {
      final String error = sendStartLeaderElection(follower.getId(), lastEntry);
      if (error != null) {
        LOG.warn("{}: Failed to sendStartLeaderElection to transferee {}",
            server.getMemberId(), follower.getId());
      } else {
        LOG.info("{}: sent StartLeaderElection to transferee {} after received AppendEntriesResponse",
            server.getMemberId(), follower.getId());
      }
    }
```

Reply:
Updated, I have swapped the two branches when checking `error`.